build/do.rq

```diff
@@ -163,6 +184,134 @@ write_readme {
+			"git",
+			"ls-remote",
+			"--tags",
+			"https://github.com/styrainc/enterprise-opa"
+		], {
+			"stdout_spec": {
+				"format": "raw",
+				"options": {
+					"raw.fs": "/",
+					"raw.rs": "[\n\r]"
+				},
+			}
+		}
+	)
+
+	{ rq.error(sprintf("\nstdout: %s\nstderr: %s\n", [eopa_tags_result.stdout, eopa_tags_result.stderr])) | eopa_tags_result.exitcode != 0 }
+
+
+	# We assume that tags and capabilities files are 1:1, but some EOPA
+	# release tags in the past did not correctly get capabilities files, so
+	# we eliminate them from consideration.
+	
+	known_bad_tags := {
+		"v0.100.5",  # tag missing capabilities file
+		"v0.100.6",  # tag missing capabilities file
+		"v0.100.7",  # tag missing capabilities file
+		"v0.49.0-5", # tag missing capabilities file
+		"v0.49.0-6", # tag missing capabilities file
+		"v0.49.0-7", # tag missing capabilities file
+		"v0.49.0-8", # tag missing capabilities file
+		"v1.15.0",   # tag missing capabilities file (misnamed v0.15.0)
+		"v1.4.1",    # tag missing capabilities file
+		"v1.5.0",    # tag missing capabilities file
+	}
+
+	eopa_tags := {t | r := eopa_tags_result.stdout[_]; t := r[2] ; not known_bad_tags[t]}
+
+	# Get a directory listing for the capabilities directory, filtering for
+	# only nonzero size files with JSON extensions. The size check is to
+	# avoid long-tail edge cases where we crashed after opening the file
+	# for writing but before committing any content.
+	eopa_caps_tree := {p: f | f := rq.tree(eopa_caps_dir, {})[p]; f.size != 0; f.ext == "json"}
+
+	# Determine which capabilities files are missing, what URL they
+	# should be fetched from, and where they should end up on disk.
+	missing_locally := {
+		{"local": p, "remote": r}
+
+		|
+
+		# construct the local path we expect the caps to exist at
+		t := eopa_tags[_]
+		p := rq.joinpath([eopa_caps_dir, sprintf("%s.json", [t])])
+
+		# eopa_caps_tree has object values, so if the path is missing,
+		# retrieving a non-object type is sufficient to check for
+		# nonexistance
+		object.get(eopa_caps_tree, p, "-") == "-"
+
+		# construct the URL to fetch the content from
+		r := rq.template("https://raw.githubusercontent.com/StyraInc/enterprise-opa/main/capabilities/{{.tag}}.json", {"tag": t})
+	}
+
+	print(sprintf("fetching %d capabilities files missing locally", [count(missing_locally)]))
+
+	# Download the capabilities from the constructed URLs.
+	new_caps := {
+		{"local": m.local, "content": c}
+		|
+		m := missing_locally[_]
+		print(sprintf("\tfetcing %s", [m.remote]))
+		resp := http.send({"url": m.remote, "method": "GET"})
+			
+		{ rq.error(sprintf("non-200 status code '%d' for URL '%s'", [resp.status_code, m.remote])) | resp.status_code != 200 }
+
+		c := resp.raw_body
+
+		# sanity check in case we got the URL wrong
+		not regex.match("404: Not Found", c)
+	}
+
+	# Commit the retrieved content to disk.
+	{
+		true
+		|
+		cap := new_caps[_]
+		rq.write(cap.content, {"format": "raw", "file_path": cap.local})
+	}
```

Comment:
hehe funny. I'm not saying that we need to change anything here, but it seems like a thing OPA could eventually learn to optimize (early-exit style): with only constant elements before the `|` in a set or object comprehension, it could stop evaluating the body when it's succeeded _once_ 🤔

Reply:
Hmm, that's a good point, I should probably move the `rq.write()` to the LHS of the comprehension. Just because it's safe _now_, I don't want this to bite us N years in the future when OPA has such an optimization. It's probably also clearer.

Reply:
39a51f0